### PR TITLE
Bug-fixes in Golang implementation

### PR DESCRIPTION
**Pull request by 1nFrastr**
Hi~ I have done some work on golang implementation, including bug-fixes and enhancements as below.

[feature]
- Check minimum version requirements of Redis-server. Require minimum version 4.0.0 for supporting HSET multiple field/value pairs and other new features.
- Complete some unit tests

[bug-fix]
- Remove unnecessary UNWATCH, go-redis(go library) will send UNWATCH automatically when Watch() function done
- Fix some obvious bugs

### CODE REVIEW

**Comment by 1nFrastr** [comment]
still working in progress :)

**Comment by 1nFrastr** [comment]
Hi,  Docter @josiahcarlson , could you please review these changes?

**Comment by josiahcarlson** [comment]
Hey @1nFrastr, thank you for your submission. I've been busy with work and family, and yes, I'll be looking at this when I get the time.

Is there something specifically time-sensitive about this?

**Comment by 1nFrastr** [comment]
Hello Dr, thanks for taking time out of your busy schedule to reply. There are no time-sensitive problems, just take your time to do this please : )

File: golang/Chapter04/model/client.go [hunk 1]
```diff
@@ -24,8 +25,7 @@ func (c *Client) ListItem(itemid, sellerid string, price float64) bool {
 		err := c.Conn.Watch(func(tx *redis.Tx) error {
 			if _, err := tx.TxPipelined(func(pipeliner redis.Pipeliner) error {
 				if !tx.SIsMember(inventory, itemid).Val() {
-					tx.Unwatch(inventory)
-					return nil
```
Comment (josiahcarlson) [inline golang/Chapter04/model/client.go:28]: Don't we need to unwatch?

File: golang/Chapter04/model/client.go [hunk 2]
```diff
@@ -55,10 +55,10 @@ func (c *Client) PurchaseItem(buyerid, itemid, sellerid string, lprice int64) bo
 	for time.Now().Unix() < end {
 		err := c.Conn.Watch(func(tx *redis.Tx) error {
 			if _, err := tx.TxPipelined(func(pipeliner redis.Pipeliner) error {
-				price := int64(pipeliner.ZScore("market:", item).Val())
+				price := int64(tx.ZScore("market:", item).Val())
 				funds, _ := tx.HGet(buyer, "funds").Int64()
 				if price != lprice || price > funds {
-					tx.Unwatch()
```
Comment (josiahcarlson) [inline golang/Chapter04/model/client.go:61]: Same question here.

File: golang/Chapter05/model/client.go [hunk 3]
```diff
@@ -154,8 +154,6 @@ func (c *Client) CleanCounters() {
 								return err
 							}
 							index--
-						} else {
-							tx.Unwatch()
```
Comment (josiahcarlson) [inline golang/Chapter05/model/client.go:158]: Is this unneeded by the Go client?

File: golang/Chapter06/model/client.go [hunk 4]
```diff
@@ -453,7 +479,7 @@ type Pack struct {
 }
 
 type Messages struct {
-	ChatId string
+	ChatId       string
```
Comment (josiahcarlson) [inline golang/Chapter06/model/client.go:482]:
```suggestion
	ChatId string
```

File: golang/Chapter07/model/client.go [hunk 5]
```diff
@@ -30,7 +30,7 @@ func Tokenize(content string) []string {
 		}
 		words.Add(match)
 	}
-	return words.Intersection(&common.STOPWORDS)
+	return words.Diff(&common.STOPWORDS)
```
Comment (josiahcarlson) [inline golang/Chapter07/model/client.go:33]: Nice find!

File: golang/Chapter04/model/client.go (hunk 1 above)
Comment (1nFrastr) [reply golang/Chapter04/model/client.go:28]: Yep, refer to the source code inside [redis/tx.go at v7.0.0-beta.5 · go-redis/redis](https://github.com/go-redis/redis/blob/v7.0.0-beta.5/tx.go) Line 66-92, we could know `WatchContext()` method always call `tx.Close()`  to perform UNWATCH operation before return.

Comment (1nFrastr) [reply golang/Chapter04/model/client.go:28]: We can demonstrate that by running the unit test [redis-in-action/redisConn_test.go at golang · 1nFrastr/redis-in-action](https://github.com/1nFrastr/redis-in-action/blob/golang/golang/redisConn/redisConn_test.go#L24-L42) while redis-cli is on MONITOR mode. 

www@ubu-server:~$ redis-cli monitor
OK
1621740375.030173 [0 10.0.2.2:13139] "select" "15"
1621740375.030620 [15 10.0.2.2:13139] "ping"
1621740375.030995 [15 10.0.2.2:13139] "info" "server"
1621740375.031619 [15 10.0.2.2:13139] "watch" "key"
1621740375.334649 [0 10.0.2.2:13142] "select" "15"
1621740375.335134 [15 10.0.2.2:13142] "incr" "key"
1621740375.335644 [15 10.0.2.2:13139] "MULTI"
1621740375.335654 [15 10.0.2.2:13139] "EXEC"
1621740375.336050 [15 10.0.2.2:13139] "unwatch"
1621740375.336449 [15 10.0.2.2:13139] "get" "key"
1621740375.336848 [15 10.0.2.2:13139] "flushdb"

**Comment by Allianzcortex** [comment]
@1nFrastr very cool. The changes in chapter02 can fix the error `cannot use item (type string) as type *redis.Z in argument to r.Conn.cmdable.ZAdd`. I'd like to see this PRs merged into master branch soon ( - :

**Review by josiahcarlson — APPROVED** [review]
Thank you for this, just getting around to the review.

**Pull request merged**